design/baremetal-operator/early-cleaning.md

```diff
@@ -0,0 +1,128 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Early out-of-band cleaning
+
+This document proposes running driver-specific out-of-band clean
+steps.
+
+## Status
+
+implementable
+
+## Summary
+
+Depending on the selected driver, BMO will run zero or more out-of-band clean
+steps to ensure the node and the BMC are in the right state. Out-of-band clean
+steps are executed via BMC (e.g. using Redfish protocol) without booting the
+ironic agent (IPA).
+
+The first drivers to receive such handling will be `idrac` and
+`idrac-virtualmedia`. For them we will run the `management.known_good_state`
+clean step that resets iDRAC and clears its job queue.
+
+It is important that these actions run before we try to boot anything on the
+target node, be it the inspection ramdisk or a 3rd party installer ISO.
+
+## Motivation
+
+iDRAC will refuse to create a job if there is already a pending job of the same
+type. To make sure we start in a coherent state, we will reset iDRAC before
+doing anything else to it.
+
+### Goals
+
+* Provide a mechanism to run out-of-band clean steps before inspection.
+* Run `known_good_state` step for iDRAC.
+
+### Non-Goals
+
+* Run any in-band (i.e. using a ramdisk) clean steps.
+* Allow a user to provide the steps to run.
+
+## Proposal
+
+## Design Details
+
+All clean steps run asynchronously in the ironic API, even out-of-band steps
+that are reasonable fast (an order of seconds). When entering the `Registering`
+state, we will check if the driver supplies any reset clean steps. If yes, we
+will start manual cleaning and wait. Otherwise we'll move on.
```

Comment:
Can we also specify the expected behavior in case of clean failure (I suppose just retry with exp backoff)?

Reply:
Isn't it what we always do with all actions?

Reply:
I wouldn't say it happens in all cases. Anyhow from a documentation/design point of view I think it could help describing the expected behaviour in such case.

Reply:
Another point: in the current implementation, the `Registering` state is a simple pass-through (it just decides the transition to `Inspecting` or `ExternallyProvisioned`) - as the registration is revalidated in almost every state before entering them. So maybe it's not the right place for a manual cleaning. As @zaneb already suggested in [1], maybe we could try to converge both the proposals into the newly `Preparing` state (even though the original idea was to have it before `Ready`, while here it seems required even before `Inspecting`)

[1] https://github.com/metal3-io/baremetal-operator/pull/292#pullrequestreview-566740018

Reply:
Preparing happens too late, we need to run before any other actions. Would you prefer we return back to the initial proposal that added a `Resetting` state?

Reply:
If this is the only place we need to run it then I support re-using the existing registering state. The fact that it has no implementation now is actually an advantage because it means we we'll only have one thing to do.

A bigger question is: how can we tell that the reset is done? The node state will be manageable both before and after, right? AIUI we'll need to store some extra state either in ironic or in the BMH. (A state transition counts, but we can't really transition to a new state as soon as we start the cleaning because it's asynchronous and may fail.)

Reply:
Fine for me then to stick in the `Registering` phase. And agree about the reset completion issue

Reply:
> A bigger question is: how can we tell that the reset is done? The node state will be manageable both before and after, right?

Right. It may be a reason to use a new state.

> we can't really transition to a new state as soon as we start the cleaning because it's asynchronous and may fail

Mmm, not sure I get it. Cleaning is asynchronous, but *starting cleaning* is a synchronous operation.

Reply:
Right, but if it fails you'd prefer the failure to be recorded in the state which is supposed to be doing the resetting, not whichever state comes after that. The latter would be confusing for users, and would make retries difficult.

The RAID patch has the same problem and they appear to be solving it with a flag in the Status. (In older versions of the patch the cleaning started in the Preparing state and any errors were recorded in the Provisioning state, which was not so bad in that case but wouldn't work as well here.)

Reply:
There were a number of options considered here, but I am not sure if we need changes to this section in the end? Please let me know and I will edit accordingly if needed. Thanks!